Autopilot/AttitudeManager/Multirotor.c

```diff
@@ -69,138 +69,101 @@ void dearmVehicle(){
+    }
+    if (getControlValue(PITCH_CONTROL_SOURCE) == RC_SOURCE){
+        (*pitchRate) = channelIn[PITCH_IN_CHANNEL - 1];
+    }
+
+    (*yawRate) = channelIn[YAW_IN_CHANNEL - 1];
 }
 
 void outputMixing(int* channelOut, int* control_Roll, int* control_Pitch, int* control_Throttle, int* control_Yaw){
-    
-    channelOut[FRONT_LEFT_MOTOR - 1] = (*control_Throttle) + (*control_Pitch) + (*control_Roll) - (*control_Yaw);// + MIN_PWM;  
-    channelOut[FRONT_RIGHT_MOTOR - 1] = (*control_Throttle) + (*control_Pitch) - (*control_Roll) + (*control_Yaw);// + MIN_PWM;  
-    channelOut[BACK_RIGHT_MOTOR - 1] = (*control_Throttle) - (*control_Pitch) - (*control_Roll) - (*control_Yaw);// + MIN_PWM; 
-    channelOut[BACK_LEFT_MOTOR - 1] = (*control_Throttle) - (*control_Pitch) + (*control_Roll) + (*control_Yaw);// + MIN_PWM;  
+# if ROTOR_TYPE == QUAD_X
+    channelOut[FRONT_LEFT_MOTOR - 1] = (*control_Throttle) + (*control_Pitch) + (*control_Roll) - (*control_Yaw);
+    channelOut[FRONT_RIGHT_MOTOR - 1] = (*control_Throttle) + (*control_Pitch) - (*control_Roll) + (*control_Yaw);
+    channelOut[BACK_RIGHT_MOTOR - 1] = (*control_Throttle) - (*control_Pitch) - (*control_Roll) - (*control_Yaw);
+    channelOut[BACK_LEFT_MOTOR - 1] = (*control_Throttle) - (*control_Pitch) + (*control_Roll) + (*control_Yaw);
+#elif ROTOR_TYPE == QUAD_P
+    channelOut[FRONT_MOTOR - 1] = (*control_Throttle) + (*control_Pitch) - (*control_Yaw);
+    channelOut[RIGHT_MOTOR - 1] = (*control_Throttle) - (*control_Roll) + (*control_Yaw);
+    channelOut[BACK_MOTOR - 1] = (*control_Throttle) - (*control_Pitch) - (*control_Yaw);
+    channelOut[LEFT_MOTOR - 1] = (*control_Throttle) + (*control_Roll) + (*control_Yaw);
+
+#endif
 }
 
 void checkLimits(int* channelOut){
-    int i = 0;
+    int i;
     for (i = 0; i < NUM_CHANNELS; i++) {
-        if (channelOut[i] > MAX_PWM) {
-            channelOut[i] = MAX_PWM;
-        } else if (channelOut[i] < MIN_PWM) {
-            channelOut[i] = MIN_PWM;
-        }
+        constrain(&(channelOut[i]), MIN_PWM, MAX_PWM);
     }
 }
 
 void highLevelControl(){
-    //If the commands come from the ground station
-    if (getControlPermission(ALTITUDE_CONTROL,ALTITUDE_CONTROL_ON,ALTITUDE_CONTROL_SHIFT) && getControlPermission(ALTITUDE_CONTROL_SOURCE,ALTITUDE_GS_SOURCE,ALTITUDE_CONTROL_SOURCE_SHIFT)) {
-        setPitchAngleSetpoint(altitudeControl(getAltitudeInput(ALTITUDE_GS_SOURCE), getAltitude()));
-        setAltitudeSetpoint(getAltitudeInput(ALTITUDE_GS_SOURCE));
-        setThrottleSetpoint(throttleControl(getAltitudeInput(ALTITUDE_GS_SOURCE),getAltitude()));
+   
+    uint8_t rollControlType = getControlValue(ROLL_CONTROL_TYPE);
+    uint8_t rollControlSource = getControlValue(ROLL_CONTROL_SOURCE);
+    if (rollControlType == ANGLE_CONTROL) {
+        setRollAngleSetpoint(getRollAngleInput(rollControlSource));
+        setRollRateSetpoint(PIDcontrol(getPID(ROLL_ANGLE), getRollAngleSetpoint() - getRoll(), MAX_ROLL_RATE / MAX_ROLL_ANGLE));
```

Comment:
I'm on the fence about the scale factor. Instead of using that, you could just multiply the output when you call the function. What it does is make it simpler to scale from an input to an output, so you wouldn't have to scale all the gains.